services/deployer/src/infra/mod.rs

```diff
@@ -0,0 +1,227 @@
+                    self.0.push(component);
+                }
+            }
+            pub(crate) fn new(options: StartOptions) -> Components {
+                let services = options.services.clone();
+                let components = services
+                    .iter()
+                    .map(Component::from)
+                    .collect::<Vec<Component>>();
+
+                let mut components = Components(components, options.clone());
+                $(components.push_except_service(stringify!($name), $name::default().into());)+
+                components.0.sort();
+                components
+            }
+            pub(crate) async fn start(&self, cfg: &ComposeTest) -> Result<(), Error> {
+                for component in &self.0 {
+                    component.start(&self.1, cfg).await?;
+                }
+                Ok(())
+            }
+        }
+
+        #[async_trait]
+        pub(crate) trait ComponentAction {
+            fn configure(&self, options: &StartOptions, cfg: Builder) -> Result<Builder, Error>;
+            async fn start(&self, options: &StartOptions, cfg: &ComposeTest) -> Result<(), Error>;
+        }
+
+        #[async_trait]
+        impl ComponentAction for Component {
+            fn configure(&self, options: &StartOptions, cfg: Builder) -> Result<Builder, Error> {
+                match self {
+                    $(Self::$name(obj) => obj.configure(options, cfg),)+
+                }
+            }
+            async fn start(&self, options: &StartOptions, cfg: &ComposeTest) -> Result<(), Error> {
+                match self {
+                    $(Self::$name(obj) => obj.start(options, cfg).await,)+
+                }
+            }
+        }
+
+        $(impl From<$name> for Component {
+            fn from(from: $name) -> Component {
+                Component::$name(from)
+            }
+        })+
+
+        $(#[derive(Default, Debug, Clone, StructOpt, Eq, PartialEq)]
+        pub(crate) struct $name {})+
+
+        impl Component {
+            fn boot_order(&self) -> u32 {
+                match self {
+                    $(Self::$name(_) => $order,)+
+                }
+            }
+        }
+
+        impl PartialOrd for Component {
+            fn partial_cmp(&self, other: &Self) -> Option<Ordering> {
+                self.boot_order().partial_cmp(&other.boot_order())
+            }
+        }
+        impl Ord for Component {
+            fn cmp(&self, other: &Self) -> Ordering {
+                self.boot_order().cmp(&other.boot_order())
+            }
+        }
+    };
+    ($($name:ident, $order:ident), +) => {
+        impl_component!($($name,$order)+);
+    };
+}
+
+// Component Name and bootstrap ordering
+// from lower to high
+impl_component! {
+    Dns,        0,
+    Nats,       0,
+    Mayastor,   1,
+    Node,       2,
+    Pool,       3,
+    Volume,     3,
+    Rest,       4,
+    Jaeger,     4,
```

Comment:
Is it possible to have Jaeger start before the control plane services? Then we could see traces as the services come up (if there were any).